In [31]:
import pandas as pd
pd.options.mode.chained_assignment = None
import tensorflow
from tensorflow import keras
import numpy as np
from sklearn.metrics import confusion_matrix,classification_report

df = pd.read_csv('customer_churn_Preprocessed_data.csv',index_col=0)

In [32]:
def ANN(x_train_ANN,x_test_ANN,y_train_ANN,y_test_ANN,loss_ANN,weights_ANN):
    model_ANN = keras.Sequential([
    keras.Input(shape=(26,)),
    keras.layers.Dense(20,activation='relu'),
    keras.layers.Dense(1,activation='sigmoid')])
    
    model_ANN.compile(optimizer='adam',loss=loss_ANN,metrics=['accuracy'])
    if weights_ANN==-1:
        model_ANN.fit(x_train_ANN,y_train_ANN,epochs=100)
    else:
        model_ANN.fit(x_train_ANN,y_train_ANN,epochs=100,class_weight=weights_ANN)
    print(model_ANN.evaluate(x_test_ANN,y_test_ANN))
    y_preds_ANN = model_ANN.predict(x_test_ANN)
    y_preds_ANN = np.round(y_preds_ANN)
    print(classification_report(y_test_ANN,y_preds_ANN))
    return y_preds_ANN

In [33]:
x = df.drop('Churn',axis='columns')
y = df['Churn']
from sklearn.model_selection import train_test_split
x_train,x_test,y_train,y_test = train_test_split(x,y,test_size=0.2,random_state=15)

In [34]:
count_class_0,count_class_1 = y_train.value_counts()
y_train.value_counts()
#imbalanced

0    4123
1    1502
Name: Churn, dtype: int64

In [37]:
df_train = pd.concat([pd.DataFrame(x_train),pd.DataFrame(y_train)],axis='columns')

df_train_class_0 = df_train[df_train.Churn==0]
df_train_class_1 = df_train[df_train.Churn==1]

df_class_1_over = df_train_class_1.sample(count_class_0,replace=True)

new_df_train = pd.concat([df_train_class_0,df_class_1_over],axis=0)

new_x_train = new_df_train.drop('Churn',axis='columns')
new_y_train = new_df_train.Churn

In [36]:
y_predicted = ANN(new_x_train,x_test,new_y_train,y_test,'binary_crossentropy',-1)

Epoch 1/100
258/258 [==============================] - 1s 1ms/step - loss: 0.5399 - accuracy: 0.7396
Epoch 2/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4977 - accuracy: 0.7601
Epoch 3/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4916 - accuracy: 0.7629
Epoch 4/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4870 - accuracy: 0.7662
Epoch 5/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4839 - accuracy: 0.7681
Epoch 6/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4819 - accuracy: 0.7670
Epoch 7/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4795 - accuracy: 0.7692
Epoch 8/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4773 - accuracy: 0.7692
Epoch 9/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4758 - accuracy: 0.7730
Epoch 10/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4745 - accura

258/258 [==============================] - 0s 1ms/step - loss: 0.4288 - accuracy: 0.7987
Epoch 82/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4284 - accuracy: 0.7986
Epoch 83/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4278 - accuracy: 0.7983
Epoch 84/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4277 - accuracy: 0.8010
Epoch 85/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4281 - accuracy: 0.8032
Epoch 86/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4273 - accuracy: 0.8005
Epoch 87/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4271 - accuracy: 0.8005
Epoch 88/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4271 - accuracy: 0.8010
Epoch 89/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4272 - accuracy: 0.7984
Epoch 90/100
258/258 [==============================] - 0s 1ms/step - loss: 0.4261 - accuracy: 

WITHOUT HANDLING:
    
                 precision    recall  f1-score   support

           0       0.83      0.87      0.85       999
           1       0.64      0.58      0.61       408

    accuracy                           0.79      1407
    macro avg      0.74      0.72      0.73      1407
    weighted avg    0.78      0.79     0.78      1407
